core/async/src/messaging.rs

```diff
@@ -1,10 +1,180 @@
+use futures::future::BoxFuture;
+use once_cell::sync::OnceCell;
 use std::sync::Arc;
 
-/// Anything that allows sending a message of type M.
-pub trait Sender<M>: Send + Sync + 'static {
+/// Trait for sending a typed message.
+pub trait CanSend<M>: Send + Sync + 'static {
     fn send(&self, message: M);
 }
 
-pub type ArcSender<M> = Arc<dyn Sender<M>>;
+/// Wraps a CanSend. This should be used to pass around an Arc<dyn CanSend<M>>, instead
+/// of spelling out that type. Using a wrapper struct allows us to define more flexible
+/// APIs.
+pub struct Sender<M: 'static> {
+    sender: Arc<dyn CanSend<M>>,
+}
```

Comment:
If you would like to limit a bit the amount of boilerplate you might want to consider sth like this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6790fc66fb505d2687c0de3f902e750d

Reply:
Ahhh I like this. It does eliminate a lot of boilerplate by getting rid of the CanSend{Async} traits. For each backing impl of Sender we would make a From impl, e.g. `impl<...> From<actix::Addr<T>> for Sender<M>` and `impl<...> From<Arc<LateBoundSender<T>>> for Sender<M>`. There are some subtle issues though:

 1. By removing the trait, I also explicitly always have to have a Sender or SenderAsync in order to send anything. That means that later on when I have:
  ```
  struct PeerManagerAdapter {
    requests: Sender<PeerManagerRequest>,
    chain_info: Sender<SetChainInfo>,
  }
  ```

there's no easy way for me to support .send() on the PeerManagerAdapter object. I'd need to refactor every case to be network_adapter.requests.send(...).

2. Another issue is the inability to specify generic bounds when implementing From. Consider how to implement From<Arc<LateBoundSender<T>>>. T needs to be something that can support sending message M. Without a dedicated trait, there would only be two options in specifying that:
  - Have T implement Fn(M). This isn't possible, because say T is actix::Addr<...>, then neither Addr nor Fn live in our crate so we cannot supply the impl.
  - Bound T with Into<Sender<M>>. The bound works, but this bound cannot be utilized to implement From<Arc<LateBoundSender<T>>>, because we don't have ownership of T and don't ever get a chance to invoke the Into<Sender<M>> that T provides. Basically the problem here is that just because T is Into<Sender<M>> doesn't mean that we can derive a Fn(M) from a &T. In fact that isn't possible anyway because &T is a reference and our Fn(M) must have 'static lifetime that'll definitely outlive the reference.

I'm happy to explore further solutions to simplify this code though. I myself have gone through several iterations; it really is quite tricky to get everything to work :(

Reply:
ok, sure. We can simplify it later. FYI I didn't mean to replace CanSend trait by Sender, but rather use Fn(M) trait directly when needed.